### PR TITLE
os_specific: zephyr: add generic platform support

### PR DESCRIPTION
**Pull request by najumon1980**
add support for uefi as well as non uefi platform for zephyr rtos.
Also add support for 32bit platforms.

### CODE REVIEW

**Comment by najumon1980** [comment]
@ acpibob could you please review and merge the PR if ok ?

**Review by tbursztyka — CHANGES_REQUESTED** [review]
I would revisit this entirely.

Have a function in zephyr side (in include/zephyr/arch/x86/acpi.h) called z_acpi_get_rsdp_phys_addr(), which one will internally call the existing code that you are basically duplicating here: less code to maintain in both projects then. (just store the physical address as well instead of the mapped one only, in zephyr side I mean). 
The code there as been working for years, it is less prone to bug and does not require to bloat the adaptation code in acpica part (which should be as minimal as possible).

**Comment by najumon1980** [comment]
> I would revisit this entirely.
> 
> Have a function in zephyr side (in include/zephyr/arch/x86/acpi.h) called z_acpi_get_rsdp_phys_addr(), which one will internally call the existing code that you are basically duplicating here: less code to maintain in both projects then. (just store the physical address as well instead of the mapped one only, in zephyr side I mean). The code there as been working for years, it is less prone to bug and does not require to bloat the adaptation code in acpica part (which should be as minimal as possible).

We are replacing the legacy acpi x86 implementation with acpica (see the draft Zephyr PR: https://github.com/zephyrproject-rtos/zephyr/pull/59917 where the x86/../acpi.c removed). With minimum acpica table initialization (AcpiInitializeTables), we can retrieve most of required tables which is design for pre-kernel initialization. 

ACPICA is already using on multiple OS and hence I believe it is more stable and also need not want worry much about acpi version as platform FW updated to newer versions.

**Comment by najumon1980** [comment]
@tbursztyka would like to know whether you got time to review my response ?

**Comment by sacdintel** [comment]
Is there any other way around to avoid hardcoding the RSDP address and can instead be passed from the Zephyr side? Much of the suggestion from @tbursztyka might help address it but I am not sure of the motive firstly behind moving on from the legacy x86 acpi implementation. Thanks!

**Comment by sacdintel** [comment]
Let me know what you would like to do. From my end, everything looks good and is ready to be merged. Since it is a major implementation change from the Zephyr side, I hope it is approved by the community and other architects also want to move ahead with this! Thanks!

**Comment by jhedberg** [comment]
> Let me know what you would like to do. From my end, everything looks good and is ready to be merged. Since it is a major implementation change from the Zephyr side, I hope it is approved by the community and other architects also want to move ahead with this! Thanks!

Since this is not very time critical, I'd suggest to wait until August, when @tbursztyka is back from vacation (I'll also be off for the next two weeks). Then we can make sure that we have proper consensus among the upstream Zephyr maintainers before proceeding with the changes.

**Comment by najumon1980** [comment]
> Let me know what you would like to do. From my end, everything looks good and is ready to be merged. Since it is a major implementation change from the Zephyr side, I hope it is approved by the community and other architects also want to move ahead with this! Thanks!

@sacdintel I will discuss Zephyr arch change in Zephyr Architecture Working Group and will update you once it got approval (although this PR will not impact/break existing functionalities).

**Comment by sacdintel** [comment]
Sounds good guys! Take your time, no rush! Thanks for the updates!

**Comment by tbursztyka** [comment]
The problem I see with this is that it bring even more zephyr API usage into an external code. So more maintenance hassle. 
Thus why I think it's better to keep it as minimal as possible. Get the rsdp pointer function from zephyr and call it here, instead of coding it here: less exposure of zephyr API, less risk of api changes affecting this code then, etc etc... not to mention the configuration option from zephyr, as well as the x86 bios stuff btw: which makes this code not zephyr arch-agnostic anymore.

**Comment by najumon1980** [comment]
> The problem I see with this is that it bring even more zephyr API usage into an external code. So more maintenance hassle. Thus why I think it's better to keep it as minimal as possible. Get the rsdp pointer function from zephyr and call it here, instead of coding it here: less exposure of zephyr API, less risk of api changes affecting this code then, etc etc... not to mention the configuration option from zephyr, as well as the x86 bios stuff btw: which makes this code not zephyr arch-agnostic anymore.

Then the entire OS abstraction layer need to move to Zephyr repo. I can move  if you think source/os_specific/service_layers/oszephyr.c is better to be part of Zephyr.

**Comment by jhedberg** [comment]
> Then the entire OS abstraction layer need to move to Zephyr repo. I can move if you think source/os_specific/service_layers/oszephyr.c is better to be part of Zephyr.

Could you elaborate a bit on that logic? I think Tomasz is just suggesting to keep the `oszephyr.c` code and set of Zephyr APIs we depend on minimal, but not getting rid of it entirely. I believe ACPICA is designed so that each supported OS has an adaption layer as part of  the ACPICA tree - not outside of it.

**Comment by tbursztyka** [comment]
Indeed, I am not proposing to ditch oszephyr.c at all. What I want is:
- a public z_arch_acpi_rsdp_lookup() function on zephyr side (it's really easy to implement since the code already exists for x86 in arch/x86/core/acpi.c) 
- then in present oszephyr.c, use that function to get rsdp

result: No x86, no EFI, no BIOS, not too much zephyr API etc... exposure into acpica.

File: source/os_specific/service_layers/oszephyr.c
Comment (tbursztyka) [inline source/os_specific/service_layers/oszephyr.c:782]: No you are still bringing CONFIG_X86_EFI from zephyr. Just get a unique function from zephyr to call, as proposed

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Review by tbursztyka — CHANGES_REQUESTED** [review]
ok now it's fine.

**Comment by sacdintel** [comment]
Is this PR ready for our review and testing? Please let me know, thanks!

**Review by jhedberg — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]

**Comment by jhedberg** [comment]
@sacdintel from our side the PR ready now i.e. feel free to do a final review & merge it if it looks good.

**Comment by najumon1980** [comment]
@sacdintel could you please review and merge if OK

**Comment by sacdintel** [comment]
Looks good, thanks guys!

**Pull request merged**
